### ISSUE TITLE
Permissions

### ISSUE THREAD

**Issue by maticzav** [graphcool/graphcool-binding#8]
Hi, so I discussed this a bit with @marktani already and he suggested that I share it here, so we can have further discussion 😄 

So, to begin with, from what I have read and from my experience I would say that it is best to separate things that might seem related in the app but are actually not. In my opinion it would be best to separate backend into `business logic` part, `database` part and `security` part. This way we can make our code a lot more manageable, as separate "systems" cannot interrupt each other and most importantly a lot easier to understand, as the logic of our app isn't mixed with security or database and vice-versa and everything is just very comprehensible.

I love how permissions are tackled in `Graphcool` services now. It's simply brilliant. I love that they are completely separate from my client and database and, best of all, that they thoroughly depend only on database, which is also completely independent. Because of this, I can test each permission separately and gradually add new ones if it need be, to suffice the requirements.  

So it actually comes down to this, in my opinion:
```yaml
  - operation: User.read
    authenticated: false
    fields:
      - name
      - school
  - operation: User.read
    authenticated: true
    fields:
      - name
      - email
      - photo
      - school
  - operation: User.read
    authenticated: true
    query: permissions/readUser.graphql
```

This is, in my opinion, the coolest thing about `permissions`; to be able to make "levels" of access - so that, referring to the example, a `User` with no authentication can access only certain fields and the one with access a bit more and if you are reading your properties you can read all of it. 

I've heard about `exists` method, but it just feels very unnatural to me. For one, because it is mixed with `resolvers`, which makes development a lot less linear, because you can't just test `queries`. (Random thought (this is actually more a `graphql-yoga` concern maybe), it would be really cool to be able to have "admin" permissions in `graphql-playground`, very similar to what `Graphcool` offers, out of the box. I think that if we separate permissions from resolvers, this could be achieved very easily) .

So actually I have already given a suggestion now 😅, why don't we implement `permissions` the way `resolvers` are implemented in `graphcool-yoga`? Something like this maybe:
```js
const server = new GraphQLServer({
  typeDefs,
  resolvers,
  permissions, // <--
  context: req => ({
    ...req,
    db: new Graphcool({
      schemaPath: DATABASE_SCHEMA_PATH,
      endpoint: process.env.GRAPHCOOL_ENDPOINT,
      secret: process.env.GRAPHCOOL_APIKEY,
    }),
  }),
  options: { port: 5000 },
})
```

Also I really love `permission` syntax, I think it would be really neat to just find a way to port it to JavaScript somehow or something similar or maybe just a "translator" (something like `schema` import).

Anyway, I hope it makes some sense, would love to discuss further on! 😄

**Comment by ph3b** [comment]
I agree! One of my main concerns about the transition to 1.0 was the fact that authorization would have to be done within each resolver. I really like this solution and hope it gets implemented!

**Comment by ForsakenHarmony** [comment]
Yeah, the permission queries were really good, I would love to get something similar back

**Comment by schickling** [comment]
I was recently having an interesting discussion with @kbrandwijk about this. We should definitely explore this. Ideally this should be implemented as an additional package on top of the Graphcool binding (and not as part of `graphql-yoga` as it should stay compatible with any GraphQL server).

We possibly need to implement some kind of middleware/concept pattern for bindings to make this work properly.

@kbrandwijk do you have some API design suggestions in mind?

**Comment by a-type** [comment]
I was also feeling less than satisfied doing authorization in the resolver layer, so I took a swipe at implementing an authorization wrapper for Prisma:

https://github.com/a-type/prisma-authorized

It's very much an experiment, but I'm fairly proud of what I was able to accomplish in a week while working outside of Prisma itself. It's basically a layer which wraps a Prisma instance and applies authorization logic against any queries run.

I'm still not entirely confident that a wrapper is the right direction to go. Along the way I definitely started to get the suspicion that this would be much easier to accomplish if authorization were somehow integrated into the Prisma server itself.

I do think, though, that if a binding-middleware kind of a solution is the way forward, you might be able to adapt my solution to run as a middleware. It's kind of clunkily emulating a middleware right now by cloning all the query/mutation functions from Prisma itself. I made a lot of assumptions about the way Prisma names things, in particular, which might benefit from having more 'insider knowledge'.

Still early stages on this project. Haven't published to npm yet. I'm thinking about tweaking some of the way memoization works internally to make it support more flexible permissions models; I'd especially like to get dynamic permission models working cleanly to open up the option to provide a permissions API similar to what's described in the OP.

**Comment by kbrandwijk** [comment]
What we basically need is to expose a method to add GraphQL middleware. This should probably be done on the `graphql-yoga` level, for the entire public schema, and/or in `graphql-binding` for individual bindings. I have had very good results using https://github.com/alekbarszczewski/graphql-add-middleware in my previous HOC framework. I believe that approach is the way forward: the ability to add middlewares to GraphQL, just like you would add middlewares to Express.

**Comment by ForsakenHarmony** [comment]
I feel like it would be more efficient for the prisma server to do permissions, because it resolves the queries including nested objects

**Comment by maticzav** [comment]
I like the Yoga-level approach. Moving permissions to Prisma would make debugging way to difficult and extremly hard to implement custom authorisation mechanisms.

**Comment by ForsakenHarmony** [comment]
How about role based permissions on prisma, but you have to manage authentication in your app

**Comment by schickling** [comment]
@ForsakenHarmony these are two different issues. See https://github.com/graphcool/prisma/issues/1768 for permission support in the Prisma layer.

@kbrandwijk I'd even go a step further and suggest a middleware library that acts on a schema level (maybe similar to [graphql-stack](https://github.com/graphcool/graphql-stack)). To make the usage easier, `graphql-yoga` could support this out of the box, but it's important that other GraphQL server libs (like express-graphql and apollo-server) can also benefit from this.

**Comment by a-type** [comment]
@schickling If I were able to hook into some kind of middleware system for a `prisma-binding` query call, based on my implementation so far, I'd probably be looking for the following changes...

* Prisma needs to support context of some sort in its queries and mutations. The middleware needs to somehow be aware of the currently authenticated user. I'm guessing it's pretty common for developers to do what I did and fetch the authenticated user and place it in the GraphQL context before query execution.
* Middleware needs to be informed of (forgive this GraphQL newbie for possibly getting some terms wrong):
  * The root type of the query (`Query` or `Mutation`)
  * The root field of the query (like `updateUser`, etc)
  * The inputs provided and their types (`{ data: 'UserUpdateInput', where: 'UserWhereUniqueInput' }`)
  * The response type (`User`)
  * A `next` function which can be called to continue down the middleware chain and eventually run the query. It returns a promise for the result of the query. The middleware itself must then either return this result or modify it in some way. Think [koajs middleware](https://github.com/koajs/koa#async-functions-node-v76).

Perhaps some of these needs could be all gathered under the umbrella of being able to do some type introspection using Prisma itself. Prisma already has access to the `typeDefs` and could definitely parse those to help middleware understand the types its dealing with.

Given those needs, middleware API might look kind of like...

```js
// new Prisma query usage
prisma.query.createUser(inputs, info, ctx); // passing in GraphQL context

// middleware example
const prismaMiddleware = async (
  inputs, // the data inputs that were passed to the query function itself
  prisma, // the Prisma instance, augmented with type introspection abilities
  context, // a Prisma context which contains info about the current query and passed in GraphQL context
  next, // runs the next middleware in the chain, resolves the end result of the query
) => {
  // optionally modify inputs to pass to next link in the chain
  const newInputs = { ...inputs, foo: 'bar' };
  // validate args before running the actual query
  const { user } = context.graphQLContext;
  // contrived authorization demo
  if (
    ['updateUser', 'deleteUser'].includes(context.rootFieldName) &&
  	inputs.where.id !== user.id
  ) {
    throw new Error('You can only modify your own User'); // query is never run
  }

  // another type of authorization, this one using inline querying to determine the result.
  // of course, in a real authorization middleware, these kinds of rules would be generated
  // from an authorization schema or loaded from a data source, not hard-coded
  if (context.rootFieldName === 'updatePost') {
    // things get tricky here... presumably, this prisma also runs the same middleware.
    // so we'd need to make this an 'admin'/'service' request that would bypass authorization.
    // in this example I'll pass a special kind of context as a hypothetical way to do that.
    const serviceContext = { user: { role: 'SERVICE', id: 'service-user' } };
    const userRelationshipResult = prisma.query.post(inputs.where, '{ user: { id } }`, serviceContext);
    if (userRelationshipResult.user.id !== user.id) {
      throw new Error('You can only modify Posts that you own');
    }
  }
  // run the rest of the chain, await the final result
  const result = await next(inputs);
  // validate the result against different rules... another minimal hardcoded demo,
  // preventing a user from ever accessing their hashed password
  if (result.password) {
    throw new Error('You do not have access to that data');
  }
  // optionally modify final result
  return {
    ...result,
    createdAt: formatDateTime(result.createdAt),
  };
}
```

Things that I'm looking for in `context` in this draft idea:

* `rootType`: `'Query'|'Mutation'`
* `rootFieldName`: `createUser`, etc.
* `inputTypes`: map of input parameter names to type names, like `{ data: 'UserCreateInput' }`
* `resultType`: string result typename, like `User`. Does not include list or required annotations (i.e. not `[User!]!`). Perhaps some middlewares would find those things useful, though. Enough reason to have `baseResultType` or something similar to delineate?
* `graphQlContext`: the original graphQL context the user passed to the query function, which came from their resolver. My authorization middleware could assume that a `user` exists on this context, or it could be configurable by the developer to look for the user in a specific place in the context.

... I think that's all I'd need based on what I've got working so far. Some of it feels a bit awkward. But still, in order to do powerful authorization, type information is needed.

**Comment by a-type** [comment]
@maticzav After doing a lot of thinking and speaking with others on the subject, I'm definitely leaning toward your approach! I think I was going way overboard trying to get granular field-level auth on the service layer. It's far easier just to make your public facing API thoughtfully, and assign scopes to users to allow them permissions at the root field level. I'm implementing that with your library now and I'm a fan so far. Nice work!

My library had to be concerned about not only whether the user could update a User (for example), but _which_ of the Users, and _which fields_, it could update. That's kind of overboard for the user-facing API. It's far simpler to only expose one mutation to update a user, and lock that mutation to only change the currently authenticated user. Instant permissions.

For an admin user I guess you could just make another mutation like `adminUpdateUser` that can accept any user's id to modify. Only users with the `adminUpdateUser` scope could use that mutation.

Kind of an 'aha!' moment when I changed my point of view and realized this is much simpler to do than I thought. I still have some ideas in mind for how my library could potentially be useful for a generic, auto-configuring admin panel suite for a Prisma app... but other than that, I'd say avoid it! 😄

**Comment by lastmjs** [comment]
Hopefully directive permissions will be considered here, as I believe they are a most elegant solution to this discussion.

https://blog.graph.cool/graphql-directive-permissions-authorization-made-easy-54c076b5368e

https://medium.com/@lastmjs/advanced-graphql-directive-permissions-with-prisma-fdee6f846044
